datafusion/physical-expr/src/schema_rewriter.rs

```diff
@@ -0,0 +1,318 @@
+    /// with the corresponding literal value from partition_values.
+    pub fn with_partition_columns(
+        mut self,
+        partition_fields: Vec<FieldRef>,
+        partition_values: Vec<ScalarValue>,
+    ) -> Self {
+        self.partition_fields = partition_fields;
+        self.partition_values = partition_values;
+        self
+    }
+
+    /// Rewrite the given physical expression to match the target schema
+    ///
+    /// This method applies the following transformations:
+    /// 1. Replaces partition column references with literal values
+    /// 2. Handles missing columns by inserting null literals
+    /// 3. Casts columns when logical and physical schemas have different types
+    pub fn rewrite(&self, expr: Arc<dyn PhysicalExpr>) -> Result<Arc<dyn PhysicalExpr>> {
+        expr.transform(|expr| self.rewrite_expr(expr)).data()
+    }
+
+    fn rewrite_expr(
+        &self,
+        expr: Arc<dyn PhysicalExpr>,
+    ) -> Result<Transformed<Arc<dyn PhysicalExpr>>> {
+        if let Some(column) = expr.as_any().downcast_ref::<Column>() {
+            return self.rewrite_column(Arc::clone(&expr), column);
+        }
+
+        Ok(Transformed::no(expr))
+    }
+
+    fn rewrite_column(
+        &self,
+        expr: Arc<dyn PhysicalExpr>,
+        column: &Column,
+    ) -> Result<Transformed<Arc<dyn PhysicalExpr>>> {
+        // Get the logical field for this column
+        let logical_field = match self.logical_file_schema.field_with_name(column.name())
+        {
+            Ok(field) => field,
+            Err(e) => {
+                // If the column is a partition field, we can use the partition value
+                if let Some(partition_value) = self.get_partition_value(column.name()) {
+                    return Ok(Transformed::yes(expressions::lit(partition_value)));
+                }
+                // If the column is not found in the logical schema and is not a partition value, return an error
+                // This should probably never be hit unless something upstream broke, but nontheless it's better
+                // for us to return a handleable error than to panic / do something unexpected.
+                return Err(e.into());
+            }
+        };
+
+        // Check if the column exists in the physical schema
+        let physical_column_index =
+            match self.physical_file_schema.index_of(column.name()) {
+                Ok(index) => index,
+                Err(_) => {
+                    if !logical_field.is_nullable() {
+                        return exec_err!(
+                        "Non-nullable column '{}' is missing from the physical schema",
+                        column.name()
+                    );
+                    }
+                    // If the column is missing from the physical schema fill it in with nulls as `SchemaAdapter` would do.
+                    // TODO: do we need to sync this with what the `SchemaAdapter` actually does?
```

Comment:
I recommend ensuring this TODO is covered by a ticket and then adding a reference to the ticket here

Reply:
For my own sake: the point here is that SchemaAdapter is a trait so it can do arbitrary things not just replace with nulls (that's what the default implementation does). It's super annoying to have to implement SchemaAdapterFactory and all of the 3 IIRC traits. I'd propose what we do here is just add hooks that follow the TreeNode APIs and make it a builder style struct.